compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaBdChain.cpp

```diff
@@ -0,0 +1,268 @@
+      AMDAIE::ConnectionOp connectionOp = maybeConnectionOp.value();
+
+      // Packet flow, do not chain BDs.
+      std::optional<AMDAIE::FlowOp> maybeFlowOp = connectionOp.getFlowOp();
+      if (!maybeFlowOp) {
+        connectionOp->emitOpError()
+            << "expected to operate on an `amdaie.flow`";
+        return WalkResult::interrupt();
+      }
+      AMDAIE::FlowOp flowOp = maybeFlowOp.value();
+      bool isPacketFlow = flowOp.getIsPacketFlow();
+      if (isPacketFlow) return WalkResult::advance();
+
+      // Repeat count > 1, do not chain BDs.
+      int32_t repeatCount = 1;
+      uint8_t numIntraAddrDim = deviceModel.getDmaProp<uint8_t>(
+          AMDAIE::AMDAIETileType::SHIMNOC, AMDAIE::AMDAIEDmaProp::NumAddrDim);
+      uint8_t numAddrDim =
+          numIntraAddrDim + deviceModel.deviceConfig.dmaNbInterDims;
+      auto sizes = npuHalfDmaCpyNdOp.getMixedSizes();
+      auto strides = npuHalfDmaCpyNdOp.getMixedStrides();
+      if (!sizes.empty() && !strides.empty()) {
+        int64_t size = getConstantIndexOrAssert(sizes[0]);
+        int64_t stride = getConstantIndexOrAssert(strides[0]);
+        if (sizes.size() == numAddrDim || stride == 0) {
+          repeatCount = size;
+        }
+      }
+      if (repeatCount > 1) return WalkResult::advance();
+
+      // Get the BD ID and tile op.
+      std::optional<AMDAIE::BdIdOp> maybeBdIdOp = npuHalfDmaCpyNdOp.getBdIdOp();
+      if (!maybeBdIdOp) {
+        npuHalfDmaCpyNdOp.emitOpError() << "must have a BD ID op";
+        return WalkResult::interrupt();
+      }
+      AMDAIE::BdIdOp bdIdOp = maybeBdIdOp.value();
+      uint32_t bdId = getConstantIndexOrAssert(bdIdOp.getValue());
+      AMDAIE::TileOp tileOp =
+          dyn_cast_if_present<AMDAIE::TileOp>(bdIdOp.getTile().getDefiningOp());
+      if (!tileOp) {
+        bdIdOp.emitOpError() << "must operate on an `amdaie.tile`";
+        return WalkResult::interrupt();
+      }
+
+      // Any duplicate BD ID from the same tile indicates that the chain
+      // cannot grow further and requires breaking to release the
+      // conflicting BD ID.
+      SmallVector<DmaChain> chainsToBreak;
+      DmaChain currDmaChain = {tileOp, connectionOp};
+      checkForChainsToBeBroken(bdId, currDmaChain, dmaChainToBdIds,
+                               chainsToBreak);
+
+      // If the chains are not to be continued, update DMA operands using
+      // the `updateChainOperands` function.
+      if (!chainsToBreak.empty()) {
+        for (auto &entry : chainsToBreak) {
+          if (failed(updateChainOperands(rewriter, dmaChainToDmaOps[entry])))
+            WalkResult::interrupt();
+          dmaChainToBdIds[entry].clear();
+          dmaChainToDmaOps[entry].clear();
+        }
+      }
+
+      dmaChainToBdIds[currDmaChain].insert(bdId);
+      // Insert at the front, as we are walking in reverse order.
+      dmaChainToDmaOps[currDmaChain].insert(
+          dmaChainToDmaOps[currDmaChain].begin(), npuHalfDmaCpyNdOp);
```

Comment:
It would be better to push into the vector and then reverse later/iterating in reverser order as inserting in the front will have linear time complexity.